src/matrix/kaldi-matrix.cc

```diff
@@ -487,11 +491,16 @@ void MatrixBase<Real>::AddMatSmat(Real alpha, const MatrixBase<Real> &A,
     KALDI_ASSERT(A.NumCols() == B.NumRows());
 
     Matrix<Real> buf(NumRows(), NumCols(), kSetZero);
-    for (int i = 0; i < NumRows(); ++i) {
-      for (int k = 0; k < B.NumRows(); ++k) {
-        for (int e = 0; e < B.Row(k).NumElements(); ++e) {
-          int j = B.Row(k).GetElement(e).first;
-          buf(i, j) += A(i, k) * B.Row(k).GetElement(e).second;
+    MatrixIndexT a_num_rows = A.NumRows();
+    MatrixIndexT b_num_rows = B.NumRows();
+    for (int i = 0; i < a_num_rows; ++i) {
+      for (int k = 0; k < b_num_rows; ++k) {
+        const SparseVector<Real> &B_row_k = B.Row(k);
+        MatrixIndexT num_elems = B_row_k.NumElements();
+        for (int e = 0; e < num_elems; ++e) {
+          const std::pair<MatrixIndexT, Real> &p = B_row_k.GetElement(e);
+          int j = p.first;
+          buf(i, j) += A(i, k) * p.second;
```

Comment:
Also, the most important optimization here is the use of cblas_xAxpy.  You should do that too.
Here, you can do it by first moving the loop over 'i' from the outside to the inside, and then, once you verify that it still works, turn the loop over 'i' into a call to cblas_xAxpy.  You might find it interesting to read the first chapter of the book 'Matrix computations' by golub et al, if you haven't already.  There is a copy on my desk (or just above the desk); you might find Yenda in our shared office and you can borrow it.